File: Go/walking-robot-simulation.go
func robotSim(commands []int, obstacles [][]int) int {
	x, y := 0, 0
	dir := 0
	obs := map[string]bool{}
	for _, ob := range obstacles {
		obs[fmt.Sprintf("%d,%d", ob[0], ob[1])] = true
	}
	dist := 0
	for _, cmd := range commands {
		switch cmd {
		case -2:
			dir = (dir - 1 + 4) % 4
		case -1:
			dir = (dir + 1 + 4) % 4
		default:
			dx, dy := 0, 0
			switch dir {
			case 0:
				dy = 1
			case 1:
				dx = 1
			case 2:
				dy = -1
			case 3:
				dx = -1
			}
			for i := 0; i < cmd; i++ {
				pos := fmt.Sprintf("%d,%d", x+dx, y+dy)
				// fmt.Println("check", pos)
				if obs[pos] {
					// fmt.Println("blocked at ", pos)
					break
				}
				x += dx
				y += dy
			}
			dist = max(dist, x*x+y*y)
		}
	}
	return dist
}

File: JavaScript/walking-robot-simulation.js
/**
 * @param {number[]} commands
 * @param {number[][]} obstacles
 * @return {number}
 */
var robotSim = function (commands, obstacles) {
    let x = 0
    let y = 0
    let dx = 0
    let dy = 1
    let max = 0
    const o = new Set()
    for (const [xi, yi] of obstacles) {
        o.add(xi * 100000 + yi)
    }
    for (const cmd of commands) {
        if (cmd === -2) {
            [dx, dy] = [-dy, dx]
        } else if (cmd === -1) {
            [dx, dy] = [dy, -dx]
        } else {
            for (let i = 0; i < cmd; i++) {
                const xi = x + dx
                const yi = y + dy
                if (o.has(xi * 100000 + yi)) {
                    break
                }
                x = xi
                y = yi
            }
            max = Math.max(max, x * x + y * y)
        }
    }
    return max
};